src/components/molecules/Visualizer/Engine/Cesium/Box/index.tsx

```diff
@@ -20,75 +35,342 @@ export type Property = {
+    activeEdgeIndex?: number; // 0 ~ 11
   };
 };
 
-// The 6 box sides defined as planes in local coordinate space.
-// ref: https://github.com/TerriaJS/terriajs/blob/cad62a45cbee98c7561625458bec3a48510f6cbc/lib/Models/BoxDrawing.ts#L161-L169
-const SIDE_PLANES: readonly CesiumPlane[] = [
-  new CesiumPlane(new Cartesian3(0, 0, 1), 0.5),
-  new CesiumPlane(new Cartesian3(0, 0, -1), 0.5),
-  new CesiumPlane(new Cartesian3(0, 1, 0), 0.5),
-  new CesiumPlane(new Cartesian3(0, -1, 0), 0.5),
-  new CesiumPlane(new Cartesian3(1, 0, 0), 0.5),
-  new CesiumPlane(new Cartesian3(-1, 0, 0), 0.5),
-];
-
-// This is used for plane ID.
-// We can use this name, for example you want to move the box when front plane is dragged.
-const SIDE_PLANE_NAMES = ["bottom", "top", "front", "back", "left", "right"];
-
-const updateTrs = (trs: TranslationRotationScale, property: Property | undefined) => {
-  const { location, height, width, length } = property?.default ?? {};
-
-  const translation = location
-    ? Cartesian3.fromDegrees(location.lng, location.lat, location.height ?? 0)
-    : undefined;
-  if (translation) {
-    Cartesian3.clone(translation, trs.translation);
-  }
-
-  // Quaternion.clone(trs.rotation, this.trs.rotation);
+const Box: React.FC<PrimitiveProps<Property>> = memo(function BoxPresenter({ layer }) {
+  const { viewer } = useCesium();
+  const ctx = useContext();
+  const { id, isVisible, property } = layer ?? {};
+  const {
+    fillColor,
+    outlineColor,
+    activeOutlineColor,
+    outlineWidth,
+    draggableOutlineColor,
+    activeDraggableOutlineColor,
+    draggableOutlineWidth,
+    height = 100,
+    width = 100,
+    length = 100,
+    pointFillColor,
+    pointOutlineColor,
+    activePointOutlineColor,
+    pointOutlineWidth,
+    axisLineColor,
+    axisLineWidth,
+    keepBoxAboveGround,
+    scalePoint = true,
+    axisLine = true,
+    cursor,
+    activeBox,
+    activeEdgeIndex,
+    activeScalePointIndex,
+  } = property?.default ?? {};
 
-  Cartesian3.clone(new Cartesian3(width || 100, length || 100, height || 100), trs.scale);
+  const [terrainHeightEstimate, setTerrainHeightEstimate] = useState(0);
+  const [trs] = useState(() =>
+    updateTrs(new TranslationRotationScale(), property, terrainHeightEstimate),
+  );
 
-  return trs;
-};
+  const isProgressSamplingTerrainHeight = useRef(false);
+  const updateTerrainHeight = useCallback(() => {
+    if (isProgressSamplingTerrainHeight.current) {
+      return;
+    }
 
-const Box: React.FC<PrimitiveProps<Property>> = memo(function BoxPresenter({ layer }) {
-  const { id, isVisible, property } = layer ?? {};
-  const { fillColor, outlineColor, outlineWidth, fill, outline } = property?.default ?? {};
+    isProgressSamplingTerrainHeight.current = true;
+    if (keepBoxAboveGround) {
+      sampleTerrainHeight(viewer.scene, trs.translation).then(v => {
+        setTerrainHeightEstimate(v ?? 0);
+        isProgressSamplingTerrainHeight.current = false;
+      });
+    }
```

Comment:
Specific to this callback: If we call this callback and `keepBoxAboveGround` is set to false, we won't calculate a new sampleTerrainHeight, and `isProgressSamplingTerrainHeight` will stay true, which means when we run this function again, it'll always return at the top and NEVER become false again, right?

Specific to this component: We use this callback in the useEffect, where we run this component IF `keepBoxAboveGround`. If we only use this function when `keepBoxAboveGround` is true, you can remove this if statement inside the function, right? That'll solve the above problem I mentioned.

Reply:
Thank you, this is my mistake. I fixed this in [fix: assigning inProgressSamplingTerrainHeight inside of IF statement](https://github.com/reearth/reearth-web/pull/357/commits/67e88b61fd94b8056fe6189788e62a4bf445571c).

Reply:
Looks good! But you missed one point I think (or maybe it is on purpose??).
Where you use `updateTerrainHeight` is always inside of a `if(!allowEnterGround){}` statement. Then, inside of `updateTerrainHeight`, you check again if `!allowEnterGround`. One of the IF statements isn't necessary, right?

Reply:
Yeees!! I removed if statement in [refactor: remove unnecessary if statement](https://github.com/reearth/reearth-web/pull/357/commits/f69880028b5eeec6726f0ad51e4ee59a4e324769).